File: src/main/java/com/example/tutorial/dsa/basics/math/PrimeNumberValidator.java
package com.example.tutorial.dsa.basics.math;

import com.example.tutorial.common.datamodel.Offer;
import com.example.tutorial.common.utils.SampleDataLoader;
import org.springframework.boot.CommandLineRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * PrimeNumberValidator
 * ----------------------------------
 * <p>This program demonstrates prime number validation and the Sieve of Eratosthenes.
 * The core problem solved here is counting primes (LeetCode #204).
 *
 * <p><b>Problem Statement:</b>
 * Given an integer n, return the number of prime numbers that are strictly less than n.
 *
 * <p><b>Real UseCase:</b> In a credit card offers system:
 * <ul>
 *   <li>Hashing algorithms - prime numbers are used in hash functions</li>
 *   <li>Cryptography basics - prime factorization for security</li>
 *   <li>Load distribution - use primes for even distribution across servers</li>
 * </ul>
 *
 * <p><b>Examples:</b>
 * <ul>
 *   <li>Input: n = 10 → Output: 4 (primes: 2, 3, 5, 7)</li>
 *   <li>Input: n = 0 → Output: 0</li>
 *   <li>Input: n = 1 → Output: 0</li>
 * </ul>
 *
 * <p><b>Company Tags:</b> Amazon
 *
 * @see <a href="https://leetcode.com/problems/count-primes/">LeetCode 204 - Count Primes</a>
 */
@Component
public class PrimeNumberValidator implements CommandLineRunner {
  public static void main(String[] args) {
    SpringApplication.run(PrimeNumberValidator.class, args);
  }

  @Override
  public void run(String... args) throws Exception {
    System.out.println("=== PrimeNumberValidator: Prime Number Demo ===\n");

    // Load credit card offers from sample data
    List<Offer> offers = SampleDataLoader.OFFERS_DTO.get();
    System.out.println("Loaded " + offers.size() + " offers from sample data.\n");

    // Demonstrate prime counting
    System.out.println("--- Counting Primes ---\n");
    int[] testCases = {10, 20, 50, 100};

    for (int n : testCases) {
      int count = countPrimes(n);
      List<Integer> primes = listPrimes(n);
      System.out.printf("Primes less than %d: %d%n", n, count);
      System.out.printf("List: %s%n%n", primes);
    }
  }

  /**
   * Counts the number of primes less than n using Sieve of Eratosthenes.
   *
   * <p><b>LOGIC:</b>
   * <ol>
   *   <li>Create a boolean array to mark composite (non-prime) numbers</li>
   *   <li>Start with 2 (first prime)</li>
   *   <li>Mark all multiples of 2 as composite</li>
   *   <li>Move to next unmarked number (3), mark its multiples</li>
   *   <li>Continue until sqrt(n)</li>
   *   <li>Count remaining unmarked numbers</li>
   * </ol>
   *
   * <p><b>Example Walkthrough:</b>
   * <pre>
   * n = 10, initially all marked as prime: [F,F,T,T,T,T,T,T,T,T]
   *                                         0 1 2 3 4 5 6 7 8 9
   *
   * Step 1: i=2, mark multiples 4,6,8: [F,F,T,T,F,T,F,T,F,T]
   * Step 2: i=3, mark multiples 6,9:   [F,F,T,T,F,T,F,T,F,F]
   *
   * Count T's from index 2: 2,3,5,7 = 4 primes
   * </pre>
   *
   * <p><b>Time Complexity: O(n log log n)</b>
   * <br>The Sieve of Eratosthenes is highly optimized for counting primes.
   * <br><i>Like crossing out numbers in a grid - very efficient.</i>
   *
   * <p><b>Space Complexity: O(n)</b>
   * <br>We use a boolean array of size n.
   * <br><i>Like having a checklist with n items.</i>
   *
   * @param n the upper limit (exclusive)
   * @return count of primes less than n
   */
  public static int countPrimes(int n) {
    if (n <= 2) {
      return 0;
    }

    // Boolean array where isPrime[i] indicates if i is prime.
    // Initially assume all numbers are prime (true).
    boolean[] isPrime = new boolean[n];
    Arrays.fill(isPrime, true);

    // 0 and 1 are not prime.
    isPrime[0] = false;
    isPrime[1] = false;

    // Sieve: mark multiples of each prime as composite.
    // We only need to check up to sqrt(n).
    for (int i = 2; i * i < n; i++) {
      if (isPrime[i]) {
        // Mark all multiples of i as not prime.
        // Start from i*i because smaller multiples were already marked.
        for (int j = i * i; j < n; j += i) {
          isPrime[j] = false;
        }
      }
    }

    // Count primes.
    int count = 0;
    for (boolean prime : isPrime) {
      if (prime) {
        count++;
      }
    }

    return count;
  }

  /**
   * Returns a list of all primes less than n.
   *
   * @param n the upper limit (exclusive)
   * @return list of primes
   */
  public static List<Integer> listPrimes(int n) {
    List<Integer> primes = new ArrayList<>();
    if (n <= 2) {
      return primes;
    }

    boolean[] isPrime = new boolean[n];
    Arrays.fill(isPrime, true);
    isPrime[0] = false;
    isPrime[1] = false;

    for (int i = 2; i * i < n; i++) {
      if (isPrime[i]) {
        for (int j = i * i; j < n; j += i) {
          isPrime[j] = false;
        }
      }
    }

    for (int i = 2; i < n; i++) {
      if (isPrime[i]) {
        primes.add(i);
      }
    }

    return primes;
  }
}

File: src/main/python/com/example/tutorial/dsa/basics/math/prime_number_validator.py
"""
PrimeNumberValidator
----------------------------------
This program demonstrates prime number validation and the Sieve of Eratosthenes.
The core problem solved here is counting primes (LeetCode #204).

Problem Statement:
    Given an integer n, return the number of prime numbers that are strictly less than n.

Real UseCase:
    In a credit card offers system:
    - Hashing algorithms - prime numbers are used in hash functions
    - Cryptography basics - prime factorization for security

Examples:
    - Input: n = 10 -> Output: 4 (primes: 2, 3, 5, 7)
    - Input: n = 0 -> Output: 0

Company Tags: Amazon

See: https://leetcode.com/problems/count-primes/
"""

import sys
import os
from typing import List

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "..", ".."))

from com.example.tutorial.common.utils.sample_data_loader import load_offers


def count_primes(n: int) -> int:
    """
    Counts the number of primes less than n using Sieve of Eratosthenes.

    LOGIC:
        1. Create a boolean array to mark composite (non-prime) numbers
        2. Start with 2 (first prime)
        3. Mark all multiples of 2 as composite
        4. Move to next unmarked number (3), mark its multiples
        5. Continue until sqrt(n)
        6. Count remaining unmarked numbers

    Example Walkthrough:
        n = 10, initially all marked as prime: [F,F,T,T,T,T,T,T,T,T]
                                                0 1 2 3 4 5 6 7 8 9

        Step 1: i=2, mark multiples 4,6,8: [F,F,T,T,F,T,F,T,F,T]
        Step 2: i=3, mark multiples 6,9:   [F,F,T,T,F,T,F,T,F,F]

        Count T's from index 2: 2,3,5,7 = 4 primes

    Time Complexity: O(n log log n)
        The Sieve of Eratosthenes is highly optimized for counting primes.
        Like crossing out numbers in a grid - very efficient.

    Space Complexity: O(n)
        We use a boolean array of size n.
        Like having a checklist with n items.

    Args:
        n: The upper limit (exclusive).

    Returns:
        Count of primes less than n.
    """
    if n <= 2:
        return 0

    # Boolean array where is_prime[i] indicates if i is prime.
    # Initially assume all numbers are prime (True).
    is_prime = [True] * n

    # 0 and 1 are not prime.
    is_prime[0] = False
    is_prime[1] = False

    # Sieve: mark multiples of each prime as composite.
    # We only need to check up to sqrt(n).
    i = 2
    while i * i < n:
        if is_prime[i]:
            # Mark all multiples of i as not prime.
            # Start from i*i because smaller multiples were already marked.
            for j in range(i * i, n, i):
                is_prime[j] = False
        i += 1

    # Count primes.
    return sum(is_prime)


def list_primes(n: int) -> List[int]:
    """
    Returns a list of all primes less than n.

    Args:
        n: The upper limit (exclusive).

    Returns:
        List of primes.
    """
    if n <= 2:
        return []

    is_prime = [True] * n
    is_prime[0] = False
    is_prime[1] = False

    i = 2
    while i * i < n:
        if is_prime[i]:
            for j in range(i * i, n, i):
                is_prime[j] = False
        i += 1

    return [i for i in range(n) if is_prime[i]]


def main():
    """Main function to demonstrate the PrimeNumberValidator."""
    print("=== PrimeNumberValidator: Prime Number Demo ===\n")

    # Load credit card offers from sample data
    offers = load_offers()
    print(f"Loaded {len(offers)} offers from sample data.\n")

    # Demonstrate prime counting
    print("--- Counting Primes ---\n")
    test_cases = [10, 20, 50, 100]

    for n in test_cases:
        count = count_primes(n)
        primes = list_primes(n)
        print(f"Primes less than {n}: {count}")
        print(f"List: {primes}\n")


if __name__ == "__main__":
    main()
